interfaces: add cifs-mount interface

datenhahn opened the pull request:
https://forum.snapcraft.io/t/interface-mount-umount-cifs-share-permission/5689

- an interface which allows mounting and unmounting of cifs shares
- currently not all permissions needed for mount/umount are included,
  it requires snaps to request the mount-observe interface as well

  - created development environment description and some basic configs/tools to
  test and reproduce my usecase here : https://github.com/datenhahn/devtools-cifs-mount-control

zyga commented on interfaces/builtin/cifs_mount_control.go:
```diff
@@ -0,0 +1,71 @@
+umount fstype=cifs /home/*/snap/@{SNAP_NAME}/common/{,**/},
+umount fstype=cifs /var/snap/@{SNAP_NAME}/common/{,**/},
+
+/run/mount/utab rw,
```
Not sure if we want this since the filesystem is not really visibly mounted from the outside.

datenhahn replied:
Without the /run/mount/utab permission the mount (or umount?) command did fail. It showed up in the apparmor log that's why I added it. But I am happy for other suggestions. I am not too deep into apparmor / snap container confinements. So any other suggestion is welcome.

zyga replied:
Well the problem is that /run/mount/utab is _shared_ with the host. So on the outside it will _seem_ from that file alone that some CIFS is mounted but when you go and look (again on the outside of snap application's point of view) they are not.

Does it fail outright if you drop this rule?

datenhahn replied:
If I remember correctly, yes. But I will retry this later and record the logs and post them here, just to be sure.

Just in case this is relevant somehow: When I executed "mount" outside of the snap the shares were not visible in the output, when I entered the snaps name space with "snap run" I was able to see them in the list.

zyga replied:
That's true (because the mount namespace shields us) but the userspace libmount library does use that file for extra bookkeeping. In some cases it can clash with what is really mounted outside so I think this needs some more investigation.

zyga commented on interfaces/builtin/cifs_mount_control.go:
```diff
@@ -0,0 +1,71 @@
+capability sys_admin,
+
+# Allow mounts to our snap-specific writable directories
+mount fstype=cifs ** -> /home/*/snap/@{SNAP_NAME}/@{SNAP_REVISION}/{,**/},
```
You can probably use @{HOMEDIRS} in place of explicit home instead. Though it doesn't matter much in snaps.

datenhahn replied:
done

zyga commented on interfaces/builtin/cifs_mount_control.go:
```diff
@@ -0,0 +1,71 @@
+capability sys_admin,
+
+# Allow mounts to our snap-specific writable directories
+mount fstype=cifs ** -> /home/*/snap/@{SNAP_NAME}/@{SNAP_REVISION}/{,**/},
```
Could you please sort this section please?

datenhahn replied:
done

zyga commented on interfaces/builtin/cifs_mount_control.go:
```diff
@@ -0,0 +1,71 @@
+mount fstype=cifs ** -> /home/*/snap/@{SNAP_NAME}/common/{,**/},
+mount fstype=cifs ** -> /var/snap/@{SNAP_NAME}/common/{,**/},
+
+umount fstype=cifs /home/*/snap/@{SNAP_NAME}/@{SNAP_REVISION}/{,**/},
```
Same

datenhahn replied:
done

zyga commented on interfaces/builtin/cifs_mount_control_test.go:
```diff
@@ -0,0 +1,113 @@
+	c.Assert(si.BaseDeclarationSlots, testutil.Contains, "cifs-mount-control")
+}
+
+func (s *CifsMountControlInterfaceSuite) TestAutoConnect(c *C) {
```
You can drop this test I think.

datenhahn replied:
done

zyga commented:
Replied to your question @datenhahn

zyga commented on interfaces/builtin/cifs_mount_control.go:
```diff
@@ -0,0 +1,71 @@
+capability sys_admin,
+
+# Allow mounts to our snap-specific writable directories
+mount fstype=cifs ** -> @{HOMEDIRS}/*/snap/@{SNAP_NAME}/@{SNAP_REVISION}/{,**/},
```
I think `**/` is equivalent to `**`.

datenhahn commented:
I also did some checks, and it seems the functionality itself is fine also without the permission, BUT it is reported as a security violation.

```
= AppArmor =
Time: Jun 18 13:27:33
Log: apparmor="DENIED" operation="open" profile="snap.smb-copier.smb-copier" name="/run/mount/utab" pid=4586 comm="umount" requested_mask="wrc" denied_mask="wrc" fsuid=0 ouid=0
File: /run/mount/utab (write)
Suggestions:
* adjust program to use $SNAP_DATA
* adjust program to use run/shm/snap.$SNAP_NAME.*
```

zyga commented:
Can you add `deny /run/mount/utab rw,` to the rules and push this change. This will keep the denial but silence the log message. Please reference this thread as a comment. This can unblock the PR in my eyes.

datenhahn commented:
I added the line and tested again. The log message vanished and functionality is still there, so from my side this workaround would be ok.

zyga commented:
Thank you for the modification. I think from my point of view the only remaining immediate thing is https://github.com/snapcore/snapd/pull/5340#discussion_r196052166. I suspect you can just drop the trailing slash after two consecutive star characters.

This will now wait for a security review from @jdstrand and on a design/naming review from @niemeyer

datenhahn commented:
@zyga done, I removed the slashes

zyga commented:
@datenhahn you need to tweak integration tests, please look at this error message:
```
Missing high-level test for interface 'cifs-mount-control'. Please add to:
* tests/lib/snaps/test-snapd-policy-app-consumer/meta/snap.yaml
* tests/lib/snaps/test-snapd-policy-app-provider-core/meta/snap.yaml (if needed)
* tests/lib/snaps/test-snapd-policy-app-provider-classic/meta/snap.yaml (if needed)
```

datenhahn commented:
@zyga I added it to the first file. I don't really know if it is needed in the other ones, but as it is similar to fuse-support and that one is not there, so I guess the cifs-mount-control also doesn't have to go in the "(if needed)" files.

What commands do I have to run, to verify that I am not missing anything else?

zyga commented:
@datenhahn usually we run `run-checks.sh` but since the branch is green I don't think you need anything else. Let's just wait for the security review from Jamie and for design review from Gustavo.

niemeyer commented on interfaces/builtin/cifs_mount_control.go:
```diff
@@ -0,0 +1,74 @@
+
+package builtin
+
+const cifsMountControlSummary = `allows to mount and unmount CIFS shares`
```
```
`allows mounting and unmounting CIFS filesystems`
```
?

datenhahn replied:
changed

niemeyer commented on interfaces/builtin/cifs_mount_control.go:
```diff
@@ -0,0 +1,74 @@
+`
+
+const cifsMountControlConnectedPlugAppArmor = `
+# Description: Allow to mount and unmount CIFS filesystems.
```
"Allow mounting and unmounting", maybe

datenhahn replied:
changed

niemeyer commented on interfaces/builtin/cifs_mount_control.go:
```diff
@@ -0,0 +1,74 @@
+# Don't log violations for this file, see discussion here:
+#  - https://github.com/snapcore/snapd/pull/5340#issuecomment-398071797
+#  - https://forum.snapcraft.io/t/namespace-awareness-of-run-mount-utab-and-libmount/5987
+deny /run/mount/utab rw,
```
The described need for "mount-observe" seems a bit curious. Why is that necessary? Is it because the tools require it, or is there another limitation that prevent the mount syscall from working at all?

zyga replied:
The use of libmount requires the effective permissions granted by `mount-observe`

datenhahn commented:
@niemeyer @zyga @jdstrand

Regarding the naming of the interface and the dependency to mount-observe I suggest you internally discuss this and then tell me in which way I should change the Pull-Request.

datenhahn commented:
@zyga is there a timeline when this change will be merged in to mainline and when approximately it would be in an ubuntu-core release? Our customer got their own brandstore and would like to drop the "--devmode" for install.

jdstrand commented on interfaces/builtin/cifs_mount_control.go:
```diff
@@ -0,0 +1,74 @@
+# Don't log violations for this file, see discussion here:
+#  - https://github.com/snapcore/snapd/pull/5340#issuecomment-398071797
+#  - https://forum.snapcraft.io/t/namespace-awareness-of-run-mount-utab-and-libmount/5987
+deny /run/mount/utab rw,
```
I don't care for the explicit denial here since deny rules take precedent over allow rules. Ie, this means if another interface came along that allowed the access, this denial would still deny it. We either need to not worry about the logged denial or extend the comment for this. I prefer the former as a reminder to do something about https://forum.snapcraft.io/t/namespace-awareness-of-run-mount-utab-and-libmount/5987/3.

datenhahn replied:
removed the explicit denial and updated the comment

jdstrand commented on interfaces/builtin/cifs_mount_control.go:
```diff
@@ -0,0 +1,74 @@
+mount fstype=cifs ** -> @{HOMEDIRS}/*/snap/@{SNAP_NAME}/@{SNAP_REVISION}/{,**},
+mount fstype=cifs ** -> @{HOMEDIRS}/*/snap/@{SNAP_NAME}/common/{,**},
+mount fstype=cifs ** -> /var/snap/@{SNAP_NAME}/@{SNAP_REVISION}/{,**},
+mount fstype=cifs ** -> /var/snap/@{SNAP_NAME}/common/{,**},
```
Using '**' means that the snap could try to mount anything from the host into SNAP_DATA or SNAP_COMMON, which would bypass intended access restrictions. That said, the fstype should cause things to fail, but a crafted mount command shipped in the snap might be used to tickle kernel bugs. What do denials look like when you try to mount cifs filesystems and you don't have these rules?

datenhahn replied:
```
= AppArmor =
Time: Jul 15 05:52:53
Log: apparmor="DENIED" operation="mount" info="failed mntpnt match" error=-13 profile="snap.smb-copier.smb-copier" name="/var/snap/smb-copier/common/shares/" pid=14381 comm="mount.cifs" fstype="cifs" srcname="//192.168.123.1/sambashare"
```

datenhahn replied:
@jdstrand what would you suggest otherwise? I would like to allow all kind of cifs options and arbitrary cifs pathes. Would it help to let the mount pathes start with // ? How would I allow all mount options and pathes starting with `//` ?

jdstrand replied:
It would look better if we could do:
```
mount fstype=cifs //** -> /var/snap/@{SNAP_NAME}/@{SNAP_REVISION}/{,**},
mount fstype=cifs //** -> /var/snap/@{SNAP_NAME}/common/{,**},
```

Can you try it locally and see if it works? I suspect however that it will not and that apparmor_parser will minimize that to '/**' internally and it won't match. If it doesn't work, simply use:

```
// source is of form '//<ip>/dir1/dirN' and is therefore not predictable so we'll
// rely on the fstype.
mount fstype=cifs -> /var/snap/@{SNAP_NAME}/@{SNAP_REVISION}/{,**},
mount fstype=cifs -> /var/snap/@{SNAP_NAME}/common/{,**},
```

jdstrand commented on interfaces/builtin/cifs_mount_control.go:
```diff
@@ -0,0 +1,74 @@
+
+# Allow mounts to our snap-specific writable directories
+mount fstype=cifs ** -> @{HOMEDIRS}/*/snap/@{SNAP_NAME}/@{SNAP_REVISION}/{,**},
+mount fstype=cifs ** -> @{HOMEDIRS}/*/snap/@{SNAP_NAME}/common/{,**},
```
That said, if there are important use cases for this functionality, please list them here.

datenhahn replied:
I removed the user home dir specific rules

datenhahn commented on interfaces/builtin/cifs_mount_control.go:
```diff
@@ -0,0 +1,74 @@
+mount fstype=cifs ** -> /var/snap/@{SNAP_NAME}/common/{,**},
+
+umount fstype=cifs @{HOMEDIRS}/*/snap/@{SNAP_NAME}/@{SNAP_REVISION}/{,**},
+umount fstype=cifs @{HOMEDIRS}/*/snap/@{SNAP_NAME}/common/{,**},
```
I removed the user home dir specific rules

datenhahn commented on interfaces/builtin/cifs_mount_control.go:
```diff
@@ -0,0 +1,74 @@
+umount fstype=cifs @{HOMEDIRS}/*/snap/@{SNAP_NAME}/@{SNAP_REVISION}/{,**},
+umount fstype=cifs @{HOMEDIRS}/*/snap/@{SNAP_NAME}/common/{,**},
+umount fstype=cifs /var/snap/@{SNAP_NAME}/@{SNAP_REVISION}/{,**},
+umount fstype=cifs /var/snap/@{SNAP_NAME}/common/{,**},
```
Changed as requested

niemeyer replied:
Isn't this a more serious problem, @jdstrand?  If the snap can unmount, it can override mounts and get access to the underlying filesystem. Not sure if there's already a security issue created by this, but it's an axis we don't generally account for when reviewing logic.

niemeyer replied:
Sorry, nevermind... I see the unmounts are restricted to $SNAP_DATA, so it won't really be a problem.

jdstrand commented on interfaces/builtin/cifs_mount_control.go:
```diff
@@ -0,0 +1,74 @@
+# decide to allow access (deny rules have precedence).
+#
+#  - https://github.com/snapcore/snapd/pull/5340#issuecomment-398071797
+#  - https://forum.snapcraft.io/t/namespace-awareness-of-run-mount-utab-and-libmount/5987`
```
You deleted the trailing '`' so this isn't syntactically correct. I suggest adding this:
```
#deny /run/mount/utab w,
`
```

(ie, add a comment and the trailing '`'`.

jdstrand commented on interfaces/builtin/cifs_mount_control_test.go:
```diff
@@ -0,0 +1,109 @@
+	spec := &apparmor.Specification{}
+	c.Assert(spec.AddConnectedPlug(s.iface, s.plug, s.slot), IsNil)
+	c.Assert(spec.SecurityTags(), DeepEquals, []string{"snap.consumer.app"})
+	c.Assert(spec.SnippetForTag("snap.consumer.app"), testutil.Contains, `/run/mount/utab`)
```
This will need to be adjusted since you removed it (please adjust to be `#deny /run/mount/utab,\\n` since I recommended a commented out rule.

jdstrand commented:
I agree with @niemeyer, please rename to 'cifs-mount'.

datenhahn commented:
@jdstrand @niemeyer 

I renamed the interface to `cifs-mount`. I wasn't able to run `./run-checks` it fails when parsing some shellscripts in the tests directories. But at least the format-check and vet phase ran through. I was able to run the cifs_mount_test.go locally (just that file in Goland IDE).

I also changed the mount command to only allow network shares ( //** ).

jdstrand commented on interfaces/builtin/cifs_mount.go:
```diff
@@ -19,33 +19,33 @@
-mount fstype=cifs ** -> /var/snap/@{SNAP_NAME}/@{SNAP_REVISION}/{,**},
-mount fstype=cifs ** -> /var/snap/@{SNAP_NAME}/common/{,**},
+mount fstype=cifs //** -> /var/snap/@{SNAP_NAME}/@{SNAP_REVISION}/{,**},
+mount fstype=cifs //** -> /var/snap/@{SNAP_NAME}/common/{,**},
```
Can you comment that this actually works for you?

datenhahn replied:
yes, I tested it with my minimal test snap and everything seemed to work fine:

https://github.com/datenhahn/devtools-cifs-mount-control

datenhahn commented:
I ran the two commandlines for running tests and both passed

jdstrand commented on interfaces/builtin/cifs_mount.go:
```diff
@@ -0,0 +1,77 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2016-2018 Canonical Ltd
```
Please adjust to '2018'

datenhahn replied:
changed

jdstrand commented on interfaces/builtin/cifs_mount.go:
```diff
@@ -0,0 +1,77 @@
+
+const cifsMountConnectedPlugSecComp = `
+# Description: Allow mount and umount syscall access.
+
```
Can you remove this newline?

datenhahn replied:
removed

jdstrand commented on interfaces/builtin/cifs_mount.go:
```diff
@@ -0,0 +1,77 @@
+# /run/mount/utab but fails. The resulting apparmor warning can be ignored. The log warning
+# was not removed via an explicit deny to not interfere with other interfaces which might
+# decide to allow access (deny rules have precedence).
+#
```
Can you remove this blank comment line?

datenhahn replied:
removed

jdstrand commented on interfaces/builtin/cifs_mount.go:
```diff
@@ -0,0 +1,77 @@
+#
+#  - https://github.com/snapcore/snapd/pull/5340#issuecomment-398071797
+#  - https://forum.snapcraft.io/t/namespace-awareness-of-run-mount-utab-and-libmount/5987
+#
```
Can you remove this blank comment line?

datenhahn replied:
removed

jdstrand commented on interfaces/builtin/cifs_mount_test.go:
```diff
@@ -0,0 +1,109 @@
+	c.Assert(spec.SecurityTags(), DeepEquals, []string{"snap.consumer.app"})
+	c.Assert(spec.SnippetForTag("snap.consumer.app"), testutil.Contains, "mount\n")
+	c.Assert(spec.SnippetForTag("snap.consumer.app"), testutil.Contains, "umount\n")
+	c.Assert(spec.SnippetForTag("snap.consumer.app"), testutil.Contains, "umount2\n")
```
Just one of these is enough.

datenhahn replied:
Not sure I understood your comment correctly. I changed it and only kept the umount2 assert (it is most specific for the cifs mount util). If one assertion is true all settings were applied successfully anyway, that was your point, right?

datenhahn commented:
@niemeyer I signed it just now and put you as "Canonical Contact" , I am not 100% sure if I chose the correct launchpad-ID (I don't have one, I sighed in with my Ubuntu-One account and I put that ID)

niemeyer commented:
@datenhahn I see it, thanks!  Changing going in...
